iceoryx_dds/test/moduletests/test_iox_to_dds.cpp

```diff
@@ -0,0 +1,382 @@
+    msg.m_subType = iox::capro::CaproMessageSubType::EVENT;
+
+    EXPECT_CALL(gw, addChannel).Times(0);
+
+    // === Test
+    gw.discover(msg);
+}
+
+TEST_F(Iceoryx2DDSGatewayTest, IgnoresServiceMessages)
+{
+    // === Setup
+    TestGateway gw{};
+    auto msg = iox::capro::CaproMessage(
+        iox::capro::CaproMessageType::OFFER,
+        {iox::capro::AnyServiceString, iox::capro::AnyInstanceString, iox::capro::AnyEventString});
+    msg.m_subType = iox::capro::CaproMessageSubType::SERVICE;
+
+    EXPECT_CALL(gw, addChannel).Times(0);
+
+    // === Test
+    gw.discover(msg);
+}
+
+TEST_F(Iceoryx2DDSGatewayTest, ChannelsAreCreatedForDiscoveredServices)
+{
+    // === Setup
+    auto testService = iox::capro::ServiceDescription({"Radar", "Front-Right", "Reflections"});
+    TestGateway gw{};
+    auto msg = iox::capro::CaproMessage(iox::capro::CaproMessageType::OFFER, testService);
+    msg.m_subType = iox::capro::CaproMessageSubType::EVENT;
+
+    EXPECT_CALL(gw, findChannel).WillOnce(Return(iox::cxx::nullopt_t()));
+    EXPECT_CALL(gw, addChannel).WillOnce(Return(createTestChannel(testService)));
+
+    // === Test
+    gw.discover(msg);
+}
+
+TEST_F(Iceoryx2DDSGatewayTest, ImmediatelySubscribesToDataFromDiscoveredServices)
+{
+    // === Setup
+    auto testService = iox::capro::ServiceDescription({"Radar", "Front-Right", "Reflections"});
+    auto mockSubscriber = createMockSubscriber(testService);
+    EXPECT_CALL(*mockSubscriber, subscribe).Times(1);
+    stageMockSubscriber(std::move(mockSubscriber));
+
+    TestGateway gw{};
+    auto msg = iox::capro::CaproMessage(iox::capro::CaproMessageType::OFFER, testService);
+    msg.m_subType = iox::capro::CaproMessageSubType::EVENT;
+
+    // Mock methods of the mock generic dds gateway base class
+    ON_CALL(gw, findChannel).WillByDefault(Return(iox::cxx::nullopt_t()));
+    ON_CALL(gw, addChannel).WillByDefault(Return(createTestChannel(testService)));
+
+    // === Test
+    gw.discover(msg);
+}
+
+TEST_F(Iceoryx2DDSGatewayTest, ImmediatelyConnectsCreatedDataWritersForDiscoveredServices)
+{
+    // === Setup
+    auto testService = iox::capro::ServiceDescription({"Radar", "Front-Right", "Reflections"});
+    auto mockWriter = createMockDataWriter(testService);
+    EXPECT_CALL(*mockWriter, connect).Times(1);
+    stageMockDataWriter(std::move(mockWriter));
+
+    TestGateway gw{};
+    auto msg = iox::capro::CaproMessage(iox::capro::CaproMessageType::OFFER, testService);
+    msg.m_subType = iox::capro::CaproMessageSubType::EVENT;
+
+    // Mock methods of the mock generic dds gateway base class
+    ON_CALL(gw, findChannel).WillByDefault(Return(iox::cxx::nullopt_t()));
+    ON_CALL(gw, addChannel).WillByDefault(Return(createTestChannel(testService)));
+
+    // === Test
+    gw.discover(msg);
+}
+
+TEST_F(Iceoryx2DDSGatewayTest, ForwardsChunkFromSubscriberToDataWriter)
```

Comment:
Zero, One, Many, Corner-Cases Rule.
Forward from a lot of subscribers to data writers - if it makes sense

Reply:
Same explanation, not stateful...